src/graphics/shader-node-registry.js

```diff
@@ -0,0 +1,88 @@
+import { ShaderGraphNode } from '../scene/materials/shader-graph-node.js';
+import { hashCode } from '../core/hash.js';
+/**
+ * @private
+ * @class
+ * @name pc.ShaderNodeRegistry
+ * @classdesc Container for all Shader Graph Core Nodes available to this application.
+ * @description Create an instance of a pc.ShaderNodeRegistry.
+ * @param {pc.Application} app - Application to attach registry to.
+ */
+function ShaderNodeRegistry(app) {
+    this.app = app;
+    this._nodeDef = { };
+    this._nodeCache = { };
+}
+
+ShaderNodeRegistry.prototype.constructor = ShaderNodeRegistry;
+
+ShaderNodeRegistry.prototype.destroy = function () {
+    this.app = null;
+};
+
+/**
+ * @private
+ * @function
+ * @name pc.ShaderNodeRegistry#register
+ * @description Register core nodes with registry.
+ * @param {object|string} coreNodeList - core node list object or JSON string
+ */
+ShaderNodeRegistry.prototype.register = function (coreNodeList) {
+    if (coreNodeList) {
+        if (typeof coreNodeList === 'string') {
+            coreNodeList = JSON.parse(coreNodeList);
+        }
+        var self = this;
+        Object.keys(coreNodeList).forEach(function (key) {
+            self._nodeDef[key] = coreNodeList[key];
+            self._nodeCache[key] = {};
+        });
+    }
+};
+
+ShaderNodeRegistry.prototype._genVariantKey = function (argTypes, options) {
+    var variantKey = 'varKey';
+
+    for (var argIndex = 0; argIndex < argTypes.length; argIndex++) {
+        variantKey += '_' + argTypes[argIndex];
+    }
+
+    if (options) {
+        var optionKeys = Object.keys(options);
+        for (var keyIndex = 0; keyIndex < optionKeys.length; keyIndex++) {
+            variantKey += '_' + optionKeys + '_' + options[optionKeys[keyIndex]];
```

Comment:
I think `+ optionKeys +` is not what you want there, as optionKeys is an array and that add all array keys there, should be `+ optionKeys[keyIndex] +` I suspect?
There is probably better way to walk this anyways, without calling `Object.keys` which creates an array, and without having to do lookup on each step `options[optionKeys[keyIndex]]`

Reply:
yes - it was a typo - refactored to fix this and be more efficient - and looks simpler